### PR TITLE
Tirspec H1RG support

### PR DESCRIPTION
**Pull request by varghesereji**
The package HxRGproc can reduce data from the instruments TANSPEC, which have H2RG Detector in spectrograph, and TIRSPEC, which is a spectrograph with H1RG Detector.

### CODE REVIEW

File: HxRGproc/reduction/instruments.py [hunk 1]
```diff
@@ -117,6 +117,33 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
+    """ Function which returns the key to sort SpecTANSPEC filename """
+    return tuple(map(int,re.search('.*-(\d+?)-(\d+?)\.fits',os.path.basename(fname)).group(1,2)))
+
+
+def fix_header_function_TIRSPEC(header,fname=None):
+   """Function to fix any missing headers needed in header"""
+   if 'CHANNELS' not in header:
+           header['CHANNELS'] = 16
+   if ('NDRITIME' not in header) or (header['NDRITIME'] == 0):
+       FrameTime = header['ITIME']  #Time taken for each readout.
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:135]: I think the header keyword for the NDR readout time is `NDRRDTM`

File: HxRGproc/reduction/instruments.py [hunk 2]
```diff
@@ -117,6 +117,33 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
+       Frame_Number = re.search('.*(\d+?)\.fits',os.path.basename(fname)).group(1)
+       time = int(Frame_Number) * FrameTime
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:137]: Unlike TANSPEC the keyword here does not reset at the beginning of an up the ramp. We should think how to subtract the first number from the up-the-ramp.

File: HxRGproc/reduction/instruments.py [hunk 3]
```diff
@@ -117,6 +117,33 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
+    return tuple(map(int,re.search('.*-(\d+?)-(\d+?)\.fits',os.path.basename(fname)).group(1,2)))
+
+
+def fix_header_function_TIRSPEC(header,fname=None):
+   """Function to fix any missing headers needed in header"""
+   if 'CHANNELS' not in header:
+           header['CHANNELS'] = 16
+   if ('NDRITIME' not in header) or (header['NDRITIME'] == 0):
+       FrameTime = header['ITIME']  #Time taken for each readout.
+       Frame_Number = re.search('.*(\d+?)\.fits',os.path.basename(fname)).group(1)
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:136]: You could use the regular expression `'.*debug-(\d+?)\.fits'` instead.

**Review by indiajoe — CHANGES_REQUESTED** [review]
Here are some suggestions. I shall add more once I test it on data.

File: HxRGproc/reduction/instruments.py [hunk 4]
```diff
@@ -117,6 +117,33 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
     return DataCube
 
+#####################################################################
+#####################################################################
+#### Functions specific to reduce TIRSPEC software data
+#####################################################################
+
+def sort_filename_key_function_TIRSPEC(fname):
+    """ Function which returns the key to sort SpecTANSPEC filename """
+    return tuple(map(int,re.search('.*-(\d+?)-(\d+?)\.fits',os.path.basename(fname)).group(1,2)))
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:127]: This Regexp should be `'.*-(\d+?)-debug-(\d+?)\.fits'`

File: HxRGproc/reduction/instruments.py [hunk 5]
```diff
@@ -151,15 +178,25 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
                    'RampidRegexp':'(.*?-\d*?)\.Z\.\d*\.fits',# Regexp to extract unique Ramp id from filename
                    'HDR_NOUTPUTS' : 'CHANNELS', # Fits header for number of output channels
                    'HDR_INTTIME' : 'NDRITIME', # Fits header for accumulated exposure time in each NDR
                    'filename_sort_func': sort_filename_key_function_SpecTANSPEC,
                    'FixHeader_func': fix_header_function_SpecTANSPEC,
                    'FixDataCube_func': fix_datacube_function_SpecTANSPEC,
                    'estimate_NoNDR_Drop_G_func': None,
-                   'ExtraHeaderCalculations_func': None}
+                   'ExtraHeaderCalculations_func': None},
+    'TIRSPEC':{'RampFilenameString':'{0}.',#Input filename structure with Ramp id substitution
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:190]: It should be a - in the end not . like in TANSPEC. `'RampFilenameString':'{0}-'`

File: HxRGproc/reduction/instruments.py [hunk 6]
```diff
@@ -151,15 +178,25 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
+                   'RampidRegexp':'(.*?-\d*?)-\.\d*\.fits',# Regexp to extract unique Ramp id from filename
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:191]: This should be `'RampidRegexp':'(.*?-\d*?)-debug-\d*\.fits'`

File: HxRGproc/reduction/instruments.py [hunk 7]
```diff
@@ -186,7 +223,12 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
                 'estimate_NoNDR_Drop_G_func':None},
     'SpecTANSPEC':{'RampFilenameString':'-{0}.Z.',#Inpui filename structure with Ramp id substitution
                    'RampidRegexp':'.*-(\d*?)\.Z\.\d*?\.fits',# Regexp to extract unique Ramp id from filename
-                   'InputSubDir' : 'fits', # Append any redundant input subdirectory to be added
+                   'InputSubDir' : '', # Append any redundant input subdirectory to be added
                    'filename_sort_func':sort_filename_key_function_SpecTANSPEC,
+                   'estimate_NoNDR_Drop_G_func':None},
+    'TIRSPEC':{'RampFilenameString':'-{0}.',#Inpui filename structure with Ramp id substitution
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:229]: It should be a - in the end not . like in TANSPEC. `'RampFilenameString':'{0}-'`

File: HxRGproc/reduction/instruments.py [hunk 8]
```diff
@@ -186,7 +223,12 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
                 'estimate_NoNDR_Drop_G_func':None},
     'SpecTANSPEC':{'RampFilenameString':'-{0}.Z.',#Inpui filename structure with Ramp id substitution
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:224]: This also should be `'RampFilenameString':'{0}.Z.'` 
No need of - in the beginning.

File: HxRGproc/reduction/instruments.py [hunk 9]
```diff
@@ -186,7 +223,12 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
...
                    'RampidRegexp':'.*-(\d*?)\.Z\.\d*?\.fits',# Regexp to extract unique Ramp id from filename
-                   'InputSubDir' : 'fits', # Append any redundant input subdirectory to be added
+                   'InputSubDir' : '', # Append any redundant input subdirectory to be added
                    'filename_sort_func':sort_filename_key_function_SpecTANSPEC,
+                   'estimate_NoNDR_Drop_G_func':None},
+    'TIRSPEC':{'RampFilenameString':'-{0}.',#Inpui filename structure with Ramp id substitution
+                   'RampidRegexp':'.*-(\d*?)-\d*?\.fits',# Regexp to extract unique Ramp id from filename
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:230]: This should be `'RampidRegexp':'(.*?-\d*?)-debug-\d*\.fits'`

File: HxRGproc/reduction/instruments.py [hunk 10]
```diff
@@ -117,6 +117,33 @@ def fix_datacube_function_SpecTANSPEC(DataCube):
     DataCube = 65536-DataCube
     return DataCube
 
+#####################################################################
+#####################################################################
+#### Functions specific to reduce TIRSPEC software data
+#####################################################################
+
+def sort_filename_key_function_TIRSPEC(fname):
+    """ Function which returns the key to sort SpecTANSPEC filename """
```
Comment (indiajoe) [inline HxRGproc/reduction/instruments.py:126]: Update docstring SpecTANSPEC -> TIRSPEC

**Review by indiajoe — CHANGES_REQUESTED** [review]
Some more edit request after testing on TIRSPEC data.

File: HxRGproc/reduction/instruments.py (hunk 1 above)
Comment (varghesereji) [reply HxRGproc/reduction/instruments.py:135]: You mean replace NDRITIME with NDRRDTM right?

Comment (indiajoe) [reply HxRGproc/reduction/instruments.py:135]: No, replace ITIME with NDRRDTM

File: HxRGproc/reduction/instruments.py (hunk 6 above)
Comment (indiajoe) [reply HxRGproc/reduction/instruments.py:191]: Is this not updated?

**Review by indiajoe — CHANGES_REQUESTED** [review]

**Review by indiajoe — APPROVED** [review]
Thanks for all the updates

**Pull request merged**